# Thông tin đồ án

## Thành viên

|   MSSV   |      Họ tên     |   GitHub  |
|:--------:|:---------------:|:---------:|
| 18120120 | Trần Chí Hào    | haotrch   |
| 18120147 | Trần Quốc Thắng | tqthangbl |

## Link thùng chứa GitHub

[https://github.com/tqthangbl/P4DS-FinalProject](https://github.com/tqthangbl/P4DS-FinalProject)

# Thu thập dữ liệu

* Nguồn dữ liệu: [Google Play Store Apps trên Kaggle](https://www.kaggle.com/lava18/google-play-store-apps)
* Chủ đề: Dữ liệu gồm 2 file .csv:
    * File googleplaystore.csv: Thông tin chi tiết của các ứng dụng trên Google Play Store
    * File googleplaystore_user_reviews.csv: Thông tin của 100 review 'liên quan nhất' của mỗi ứng dụng.
* License của dữ liệu:\
    Trong file license.txt có ghi như sau:
    > This work is licensed under the Creative Commons Attribution 3.0 Unported License. To view a copy of this license, visit http://creativecommons.org/licenses/by/3.0/.

    Trong link đính kèm lại ghi như sau:
    > You are free to:\
    **Share** — copy and redistribute the material in any medium or format\
    **Adapt** — remix, transform, and build upon the material for any purpose, even commercially.
    
    Như vậy, dữ liệu này được phép sử dụng.    
* Cách thu thập dữ liệu: Dữ liệu được thu thập từ Google Play Store. Theo nguồn dữ liệu, việc thu thập dữ liệu gặp khó khăn khi Google Play Store sử dụng các kỹ thuật hiện đại phức tạp (như tải trang động) sử dụng JQuery. Version 1 của dữ liệu được đưa lên vào ngày 05/09/2018, bản hiện tại là Version 6 được cập nhật ngày 03/02/2019.

# Khám phá dữ liệu

## Import các thư viện cần thiết

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

## Đọc dữ liệu từ file csv vào dataframe

In [2]:
app_df = pd.read_csv('Data/googleplaystore.csv')
app_df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10841 rows x 13 columns]

In [3]:
review_df = pd.read_csv('Data/googleplaystore_user_reviews.csv')
review_df

App  \
0            10 Best Foods for You   
1            10 Best Foods for You   
2            10 Best Foods for You   
3            10 Best Foods for You   
4            10 Best Foods for You   
...                            ...   
64290  Houzz Interior Design Ideas   
64291  Houzz Interior Design Ideas   
64292  Houzz Interior Design Ideas   
64293  Houzz Interior Design Ideas   
64294  Houzz Interior Design Ideas   

                                       Translated_Review Sentiment  \
0      I like eat delicious food. That's I'm cooking ...  Positive   
1        This help eating healthy exercise regular basis  Positive   
2                                                    NaN       NaN   
3             Works great especially going grocery store  Positive   
4                                           Best idea us  Positive   
...                                                  ...       ...   
64290                                                NaN       NaN   
64291                                                NaN       NaN   
64292                                                NaN       NaN   
64293                                                NaN       NaN   
64294                                                NaN       NaN   

       Sentiment_Polarity  Sentiment_Subjectivity  
0                    1.00                0.533333  
1                    0.25                0.288462  
2                     NaN                     NaN  
3                    0.40                0.875000  
4                    1.00                0.300000  
...                   ...                     ...  
64290                 NaN                     NaN  
64291                 NaN                     NaN  
64292                 NaN                     NaN  
64293                 NaN                     NaN  
64294                 NaN                     NaN  

[64295 rows x 5 columns]

## Tiền xử lý và khám phá dữ liệu bảng `app_df`

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [4]:
def get_app_df_shape():
    app_num_rows, app_num_cols = app_df.shape
    print('Number of rows in app_df:', app_num_rows)
    print('Number of columns in app_df:', app_num_cols)
    
get_app_df_shape()

Number of rows in app_df: 10841
Number of columns in app_df: 13


Bảng `app_df` có 10841 dòng và 13 cột.

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Theo mô tả dữ liệu, một dòng cho biết thông tin của một ứng dụng. Theo quan sát sơ bộ ban đầu, có lẽ không gặp vấn đề các dòng có ý nghĩa khác nhau. Tuy nhiên trong lúc khám phá dữ liệu (tại phần 2.3.5.1, khi chuyển dtype của cột `Rating` sang float), có phát hiện một dòng "bất thường". Ta sẽ tiền xử lý dòng này.

Theo quan sát sơ bộ cũng như mô tả dữ liệu, có thể cho rằng cột `Reviews` có dtype là int. Thử chuyển dtype cho cột `Reviews` và bắt lỗi: 

In [5]:
try:
    app_df['Reviews'] = app_df['Reviews'].astype(int)
except Exception as e:
    print(e)

invalid literal for int() with base 10: '3.0M'


Như vậy trong cột `Reviews` tồn tại một giá trị là `'3.0M'`, nên không thể chuyển dtype cho cột này được. Trước hết ta quan sát dòng có chứa giá trị này:

In [6]:
app_df[app_df['Reviews'] == '3.0M']

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

Theo quan sát và qua quá trình khám phá dữ liệu ở các phần sau, ta thấy rằng dòng này bị thiếu giá trị của cột `Category`, khiến cho các giá trị ở các cột sau bị dịch sang bên trái một cột, làm dòng bị sai lệch ý nghĩa. Vì vậy ta sẽ sửa lại các giá trị thuộc dòng này:

In [7]:
index_change_row = app_df.loc[app_df['Reviews'] == '3.0M'].index

app_df.loc[index_change_row] = \
    'Life Made WI-Fi Touchscreen Photo Frame',\
    np.nan,\
    1.9,\
    '19',\
    '3.0M',\
    '1,000+',\
    'Free',\
    '0',\
    'Everyone',\
    np.nan,\
    'February 11, 2018',\
    '1.0.19',\
    '4.0 and up'

app_df.loc[index_change_row]

App Category  Rating Reviews  Size  \
10472  Life Made WI-Fi Touchscreen Photo Frame      NaN     1.9      19  3.0M   

      Installs  Type Price Content Rating Genres       Last Updated  \
10472   1,000+  Free     0       Everyone    NaN  February 11, 2018   

      Current Ver Android Ver  
10472      1.0.19  4.0 and up

### Dữ liệu có các dòng bị lặp không?

Xét dữ liệu có tồn tại các dòng bị lặp hay không:

In [8]:
app_df.duplicated().sum() > 0

True

Do dữ liệu tồn tại các dòng bị lặp, vì vậy ta cần loại bỏ các dòng trùng lặp.

In [9]:
app_df = app_df.drop_duplicates()
app_df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
1         3.9     967                 14M     500,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                     Genres      Last Updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
1                   2.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[10358 rows x 13 columns]

In [10]:
get_app_df_shape()

Number of rows in app_df: 10358
Number of columns in app_df: 13


Sau khi loại các dòng trùng lặp, bảng `app_df` có 10358 dòng và 13 cột.

Xem xét chi tiết hơn, theo mô tả dữ liệu, cột `App` có tất cả 9660 giá trị khác nhau. Do đó ta muốn kiểm tra rằng liệu những dòng có tên ứng dụng trùng nhau có phải là các ứng dụng khác nhau hay không.\
Xét một số dòng có giá trị cột `App` trùng nhau (bằng cách sắp xếp các dòng tăng dần theo cột `App` và giảm dần theo cột `Reviews`, xem 60 dòng đầu và 60 dòng cuối của bảng sau khi sắp xếp):

In [11]:
app_df[app_df.duplicated(subset='App', keep=False)]\
.sort_values(['App', 'Reviews'], ascending=[True, False]).head(60)

App             Category  \
3083                             365Scores - Live Scores               SPORTS   
5415                             365Scores - Live Scores               SPORTS   
1871                                         8 Ball Pool                 GAME   
1970                                         8 Ball Pool                 GAME   
1844                                         8 Ball Pool                 GAME   
1755                                         8 Ball Pool                 GAME   
1703                                         8 Ball Pool                 GAME   
1675                                         8 Ball Pool                 GAME   
3953                                         8 Ball Pool               SPORTS   
662                      95Live -SG#1 Live Streaming App               DATING   
559                      95Live -SG#1 Live Streaming App               DATING   
4991               A&E - Watch Full Episodes of TV Shows               FAMILY   
920                A&E - Watch Full Episodes of TV Shows        ENTERTAINMENT   
3799                       AC - Tips & News for Android™   NEWS_AND_MAGAZINES   
4928                       AC - Tips & News for Android™   NEWS_AND_MAGAZINES   
3791                           AP Mobile - Breaking News   NEWS_AND_MAGAZINES   
5448                           AP Mobile - Breaking News   NEWS_AND_MAGAZINES   
2800                                                ASOS             SHOPPING   
2771                                                ASOS             SHOPPING   
3628   AccuWeather: Daily Forecast & Live Weather Rep...              WEATHER   
5512   AccuWeather: Daily Forecast & Live Weather Rep...              WEATHER   
1143                        Acorns - Invest Spare Change              FINANCE   
1095                        Acorns - Invest Spare Change              FINANCE   
3524                                Adobe Acrobat Reader         PRODUCTIVITY   
3452                                Adobe Acrobat Reader         PRODUCTIVITY   
3895                                Adobe Acrobat Reader         PRODUCTIVITY   
2936   Adobe Photoshop Express:Photo Editor Collage M...          PHOTOGRAPHY   
2814   Adobe Photoshop Express:Photo Editor Collage M...          PHOTOGRAPHY   
1947                                             Agar.io                 GAME   
3880                                             Agar.io                 GAME   
3215                                              Airbnb     TRAVEL_AND_LOCAL   
3116                                              Airbnb     TRAVEL_AND_LOCAL   
2790        AliExpress - Smarter Shopping, Better Living             SHOPPING   
2660        AliExpress - Smarter Shopping, Better Living             SHOPPING   
2756        AliExpress - Smarter Shopping, Better Living             SHOPPING   
4099        AliExpress - Smarter Shopping, Better Living             SHOPPING   
2982                 All Football - Latest News & Videos               SPORTS   
5075                 All Football - Latest News & Videos               SPORTS   
10171                                All Social Networks               SOCIAL   
2549                                 All Social Networks               SOCIAL   
1933                                    Alto's Adventure                 GAME   
3936                                    Alto's Adventure                 GAME   
4083                                       Amazon Kindle  BOOKS_AND_REFERENCE   
143                                        Amazon Kindle  BOOKS_AND_REFERENCE   
2664                                     Amazon Shopping             SHOPPING   
2763                                     Amazon Shopping             SHOPPING   
3922                                     Amazon Shopping             SHOPPING   
2654                                  Amazon for Tablets             SHOPPING   
5353                                  Amazon for Tablets             SHOPPING   
315

In [12]:
app_df[app_df.duplicated(subset='App', keep=False)]\
.sort_values(['App', 'Reviews'], ascending=[True, False]).tail(60)

App         Category  \
4821                             Zello PTT Walkie Talkie           SOCIAL   
1255    Zomato - Restaurant Finder and Food Delivery App   FOOD_AND_DRINK   
1237    Zomato - Restaurant Finder and Food Delivery App   FOOD_AND_DRINK   
1982                                     Zombie Catchers             GAME   
1898                                     Zombie Catchers             GAME   
1924                                     Zombie Catchers             GAME   
1727                                     Zombie Catchers             GAME   
1693                                     Zombie Catchers             GAME   
4244                                     Zombie Catchers             GAME   
1948                                  Zombie Hunter King             GAME   
1874                                  Zombie Hunter King             GAME   
1728                                  Zombie Hunter King             GAME   
1669                                  Zombie Hunter King             GAME   
1914                                      Zombie Tsunami             GAME   
2008                                      Zombie Tsunami             GAME   
1730                                      Zombie Tsunami             GAME   
1685                                      Zombie Tsunami             GAME   
3940                                      Zombie Tsunami             GAME   
7077                      Zoosk Dating App: Meet Singles           DATING   
482                       Zoosk Dating App: Meet Singles           DATING   
3445                         ai.type Free Emoji Keyboard  PERSONALIZATION   
5863                         ai.type Free Emoji Keyboard  PERSONALIZATION   
9635    busuu: Learn Languages - Spanish, English & More           FAMILY   
835     busuu: Learn Languages - Spanish, English & More        EDUCATION   
808     busuu: Learn Languages - Spanish, English & More        EDUCATION   
2662   eBay: Buy & Sell this Summer - Discover Deals ...         SHOPPING   
4068   eBay: Buy & Sell this Summer - Discover Deals ...         SHOPPING   
743          edX - Online Courses by Harvard, MIT & more        EDUCATION   
789          edX - Online Courses by Harvard, MIT & more        EDUCATION   
389                            free video calls and chat    COMMUNICATION   
4224                           free video calls and chat    COMMUNICATION   
383                        imo free video calls and chat    COMMUNICATION   
347                        imo free video calls and chat    COMMUNICATION   
2663    letgo: Buy & Sell Used Stuff, Cars & Real Estate         SHOPPING   
4322    letgo: Buy & Sell Used Stuff, Cars & Real Estate         SHOPPING   
673    muzmatch: Muslim & Arab Singles, Marriage & Da...           DATING   
525    muzmatch: Muslim & Arab Singles, Marriage & Da...           DATING   
3492                                              myAT&T     PRODUCTIVITY   
4593                                              myAT&T     PRODUCTIVITY   
2331   mySugr: the blood sugar tracker made just for you          MEDICAL   
6194   mySugr: the blood sugar tracker made just for you          MEDICAL   
2595              ooVoo Video Calls, Messaging & Stories           SOCIAL   
2618              ooVoo Video Calls, Messaging & Stories           SOCIAL   
3334                             osmino Wi-Fi: free WiFi            TOOLS   
10473                            osmino Wi-Fi: free WiFi            TOOLS   
5664       realestate.com.au - Buy, Rent & Sell Property   HOUSE_AND_HOME   
1468       realestate.com.au - Buy, Rent & Sell Property   HOUSE_AND_HOME   
1891                                          slither.io             GAME   
1735                                          slither.io             GAME   
1757                                          slither.io             GAME   
1659                                          slither.io             GAME   
4227                                          sli

Theo quan sát sơ bộ, các dòng trùng nhau tại giá trị cột `App` đều có các thông tin giống nhau ở nhiều cột khác, nhưng khác nhau tối thiểu tại giá trị ở cột `Reviews`. Do đó với mỗi tên ứng dung, ta sẽ giữ lại dòng có giá trị ở cột `Reviews` cao nhất (có lẽ càng nhiều reviews sẽ cho thấy rằng dòng dữ liệu được thu thập ở thời gian mới nhất), và loại bỏ các dòng còn lại.

In [13]:
app_df = app_df.sort_values(['App', 'Reviews'], ascending=[True, False])\
                .drop_duplicates(subset='App').sort_index()
app_df

App             Category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
5                             Paper flowers instructions       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       Rating Reviews                Size     Installs  Type Price  \
0         4.1     159                 19M      10,000+  Free     0   
2         4.7   87510                8.7M   5,000,000+  Free     0   
3         4.5  215644                 25M  50,000,000+  Free     0   
4         4.3     967                2.8M     100,000+  Free     0   
5         4.4     167                5.6M      50,000+  Free     0   
...       ...     ...                 ...          ...   ...   ...   
10836     4.5      38                 53M       5,000+  Free     0   
10837     5.0       4                3.6M         100+  Free     0   
10838     NaN       3                9.5M       1,000+  Free     0   
10839     4.5     114  Varies with device       1,000+  Free     0   
10840     4.5  398307                 19M  10,000,000+  Free     0   

      Content Rating                   Genres      Last Updated  \
0           Everyone             Art & Design   January 7, 2018   
2           Everyone             Art & Design    August 1, 2018   
3               Teen             Art & Design      June 8, 2018   
4           Everyone  Art & Design;Creativity     June 20, 2018   
5           Everyone             Art & Design    March 26, 2017   
...              ...                      ...               ...   
10836       Everyone                Education     July 25, 2017   
10837       Everyone                Education      July 6, 2018   
10838       Everyone                  Medical  January 20, 2017   
10839     Mature 17+        Books & Reference  January 19, 2015   
10840       Everyone                Lifestyle     July 25, 2018   

              Current Ver         Android Ver  
0                   1.0.0        4.0.3 and up  
2                   1.2.4        4.0.3 and up  
3      Varies with device          4.2 and up  
4                     1.1          4.4 and up  
5                     1.0          2.3 and up  
...                   ...                 ...  
10836                1.48          4.1 and up  
10837                 1.0          4.1 and up  
10838                 1.0          2.2 and up  
10839  Varies with device  Varies with device  
10840  Varies with device  Varies with device  

[9660 rows x 13 columns]

In [14]:
get_app_df_shape()

Number of rows in app_df: 9660
Number of columns in app_df: 13


Sau khi loại bỏ các dòng ở bước trên, bảng `app_df` có 9660 dòng và 13 cột.

### Mỗi cột có ý nghĩa gì?

Theo mô tả dữ liệu, các cột có ý nghĩa như sau:
* **App**: Tên ứng dụng
* **Category**: Hạng mục chính mà ứng dụng thuộc về
* **Rating**: Mức đánh giá tổng thể của ứng dụng (tại thời điểm thu thập dữ liệu)
* **Reviews**: Số lượng người đánh giá ứng dụng (tại thời điểm thu thập dữ liệu)
* **Size**: Kích thước ứng dụng (tại thời điểm thu thập dữ liệu)
* **Installs**: Số lượng người tải xuống / cài đặt ứng dụng (tại thời điểm thu thập dữ liệu)
* **Type**: Kiểu ứng dụng miễn phí (Free) hay trả phí (Paid)
* **Price**: Số tiền trả phí của ứng dụng (tại thời điểm thu thập dữ liệu)
* **Content Rating**: Nhóm độ tuổi mà ứng dụng nhắm đến
* **Genres**: Các thể loại mà ứng dụng thuộc về (có thể bao gồm cả hạng mục chính)
* **Last Updated**: Ngày cuối cùng mà ứng dụng được cập nhật trên Store (tại thời điểm thu thập dữ liệu)
* **Current Ver**: Phiên bản hiện tại của ứng dụng có mặt trên Store (tại thời điểm thu thập dữ liệu)
* **Android Ver**: Phiên bản Android yêu cầu tối thiểu (tại thời điểm thu thập dữ liệu)

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

Xét dtype hiện tại của các cột dữ liệu:

In [15]:
app_df.dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

#### Tiền xử lý cột `Reviews`

Sau bước tiền xử lý và khám phá dữ liệu tại mục 2.3.3, ta đã có thể thực hiện chuyển đổi dtype của cột `Reviews` sang int:

In [16]:
app_df['Reviews'] = app_df['Reviews'].astype(int)
app_df['Reviews']

0           159
2         87510
3        215644
4           967
5           167
          ...  
10836        38
10837         4
10838         3
10839       114
10840    398307
Name: Reviews, Length: 9660, dtype: int64

#### Tiền xử lý cột `Size`

Quan sát các giá trị khác nhau trong cột `Size`:

In [17]:
np.array(sorted(app_df['Size'].unique()))

array(['1.0M', '1.1M', '1.2M', '1.3M', '1.4M', '1.5M', '1.6M', '1.7M',
       '1.8M', '1.9M', '10.0M', '100M', '1020k', '103k', '108k', '10M',
       '116k', '118k', '11M', '11k', '121k', '122k', '12M', '13M', '141k',
       '143k', '144k', '14M', '14k', '153k', '154k', '157k', '15M',
       '160k', '161k', '164k', '169k', '16M', '170k', '172k', '173k',
       '175k', '176k', '17M', '17k', '186k', '18M', '18k', '190k', '191k',
       '192k', '193k', '196k', '19M', '2.0M', '2.1M', '2.2M', '2.3M',
       '2.4M', '2.5M', '2.6M', '2.7M', '2.8M', '2.9M', '200k', '201k',
       '203k', '206k', '208k', '209k', '20M', '20k', '210k', '219k',
       '21M', '220k', '221k', '222k', '226k', '228k', '22M', '232k',
       '234k', '237k', '238k', '239k', '23M', '23k', '240k', '241k',
       '243k', '245k', '246k', '24M', '24k', '251k', '253k', '257k',
       '259k', '25M', '25k', '266k', '269k', '26M', '26k', '270k', '27M',
       '27k', '280k', '283k', '288k', '28M', '28k', '292k', '293k', '29M',
   

Ta sẽ chuyển đổi các giá trị trong cột `Size` về dạng số, với đơn vị là KB. Với các giá trị `'Varies with device'`, ta sẽ gán giá trị `nan`, để cột `Size` được xét là một cột numerical.

In [18]:
def to_kb(string):
    if string == 'Varies with device':
        return np.nan
    num = float(string[:-1])
    coef = 1 #Nếu chuỗi kết thúc bằng kí tự 'k', chỉ cần lấy phần số trong chuỗi
    if string[-1] == 'M':
        coef *= 1024 #Nếu chuỗi kết thúc bằng kí tự 'M', lấy phần số trong chuỗi nhân 1024
    return float(num * coef)

app_df['Size'] = app_df['Size'].apply(to_kb)
app_df['Size']

0        19456.0
2         8908.8
3        25600.0
4         2867.2
5         5734.4
          ...   
10836    54272.0
10837     3686.4
10838     9728.0
10839        NaN
10840    19456.0
Name: Size, Length: 9660, dtype: float64

#### Tiền xử lý cột `Installs`

Quan sát các giá trị khác nhau trong cột `Installs`:

In [19]:
np.array(sorted(app_df['Installs'].unique()))

array(['0', '0+', '1+', '1,000+', '1,000,000+', '1,000,000,000+', '10+',
       '10,000+', '10,000,000+', '100+', '100,000+', '100,000,000+', '5+',
       '5,000+', '5,000,000+', '50+', '50,000+', '50,000,000+', '500+',
       '500,000+', '500,000,000+'], dtype='<U14')

Ta thấy rằng cột `Installs` thuộc loại cột categorical có thứ tự, nên ta sẽ định nghĩa thứ tự các giá trị này và áp dụng cho cột `Installs`.

In [20]:
installs_cate_list = ['0', '0+', '1+', '5+', '10+', '50+',
                     '100+', '500+', '1,000+', '5,000+', '10,000+', '50,000+',
                     '100,000+', '500,000+', '1,000,000+', '5,000,000+', '10,000,000+', '50,000,000+',
                     '100,000,000+', '500,000,000+', '1,000,000,000+']
app_df['Installs'] = pd.Categorical(app_df['Installs'], categories=installs_cate_list, ordered=True)
app_df['Installs']

0            10,000+
2         5,000,000+
3        50,000,000+
4           100,000+
5            50,000+
            ...     
10836         5,000+
10837           100+
10838         1,000+
10839         1,000+
10840    10,000,000+
Name: Installs, Length: 9660, dtype: category
Categories (21, object): ['0' < '0+' < '1+' < '5+' ... '50,000,000+' < '100,000,000+' < '500,000,000+' < '1,000,000,000+']

#### Tiền xử lý cột `Type`

Quan sát tập các giá trị khác nhau của cột `Type`:

In [21]:
app_df['Type'].unique()

array(['Free', 'Paid', nan], dtype=object)

Ta thấy cột có tồn tại giá trị `nan`. Xét các dòng có giá trị `nan` tại cột `Type`:

In [22]:
index_check_row = app_df[app_df['Type'].isna()].index
app_df.loc[index_check_row]

App Category  Rating  Reviews  Size Installs Type  \
9148  Command & Conquer: Rivals   FAMILY     NaN        0   NaN        0  NaN   

     Price Content Rating    Genres   Last Updated         Current Ver  \
9148     0   Everyone 10+  Strategy  June 28, 2018  Varies with device   

             Android Ver  
9148  Varies with device

Nhận xét từ mô tả dữ liệu, nếu cột `Type` có giá trị là `'Free'` thì cột `Price` có giá trị là `0` và ngược lại. Do đó ta sẽ điền vào giá trị `Free` vào vị trí chứa giá trị `nan` này.

In [23]:
app_df['Type'].fillna('Free', inplace=True)
app_df.loc[index_check_row]

App Category  Rating  Reviews  Size Installs  \
9148  Command & Conquer: Rivals   FAMILY     NaN        0   NaN        0   

      Type Price Content Rating    Genres   Last Updated         Current Ver  \
9148  Free     0   Everyone 10+  Strategy  June 28, 2018  Varies with device   

             Android Ver  
9148  Varies with device

#### Tiền xử lý cột `Price`

Quan sát tập các giá trị khác nhau của cột `Price`:

In [24]:
app_df['Price'].unique()

array(['0', '$4.99', '$3.99', '$6.99', '$1.49', '$2.99', '$7.99', '$5.99',
       '$3.49', '$1.99', '$9.99', '$7.49', '$0.99', '$9.00', '$5.49',
       '$10.00', '$11.99', '$79.99', '$16.99', '$14.99', '$1.00',
       '$29.99', '$12.99', '$2.49', '$24.99', '$10.99', '$1.50', '$19.99',
       '$15.99', '$33.99', '$74.99', '$39.99', '$3.95', '$4.49', '$1.70',
       '$8.99', '$2.00', '$3.88', '$25.99', '$399.99', '$17.99',
       '$400.00', '$3.02', '$1.76', '$4.84', '$4.77', '$1.61', '$2.50',
       '$1.59', '$6.49', '$1.29', '$5.00', '$13.99', '$299.99', '$379.99',
       '$37.99', '$18.99', '$389.99', '$19.90', '$8.49', '$1.75',
       '$14.00', '$4.85', '$46.99', '$109.99', '$154.99', '$3.08',
       '$2.59', '$4.80', '$1.96', '$19.40', '$3.90', '$4.59', '$15.46',
       '$3.04', '$4.29', '$2.60', '$3.28', '$4.60', '$28.99', '$2.95',
       '$2.90', '$1.97', '$200.00', '$89.99', '$2.56', '$30.99', '$3.61',
       '$394.99', '$1.26', '$1.20', '$1.04'], dtype=object)

Ta sẽ giữ lại phần số trong các chuỗi giá trị, sau đó chuyển kiểu dữ liệu của cột `Price` về float:

In [25]:
app_df['Price'] = app_df['Price'].str.extract(r'\$*(.+)', expand=False).astype(float)
app_df['Price']

0        0.0
2        0.0
3        0.0
4        0.0
5        0.0
        ... 
10836    0.0
10837    0.0
10838    0.0
10839    0.0
10840    0.0
Name: Price, Length: 9660, dtype: float64

In [26]:
app_df['Price'].unique()

array([  0.  ,   4.99,   3.99,   6.99,   1.49,   2.99,   7.99,   5.99,
         3.49,   1.99,   9.99,   7.49,   0.99,   9.  ,   5.49,  10.  ,
        11.99,  79.99,  16.99,  14.99,   1.  ,  29.99,  12.99,   2.49,
        24.99,  10.99,   1.5 ,  19.99,  15.99,  33.99,  74.99,  39.99,
         3.95,   4.49,   1.7 ,   8.99,   2.  ,   3.88,  25.99, 399.99,
        17.99, 400.  ,   3.02,   1.76,   4.84,   4.77,   1.61,   2.5 ,
         1.59,   6.49,   1.29,   5.  ,  13.99, 299.99, 379.99,  37.99,
        18.99, 389.99,  19.9 ,   8.49,   1.75,  14.  ,   4.85,  46.99,
       109.99, 154.99,   3.08,   2.59,   4.8 ,   1.96,  19.4 ,   3.9 ,
         4.59,  15.46,   3.04,   4.29,   2.6 ,   3.28,   4.6 ,  28.99,
         2.95,   2.9 ,   1.97, 200.  ,  89.99,   2.56,  30.99,   3.61,
       394.99,   1.26,   1.2 ,   1.04])

#### Làm rõ ý nghĩa cột `Type` và cột `Price`

Theo mô tả dữ liệu, cột `Type` và cột `Price` có ý nghĩa như sau:
* **Type**: Kiểu ứng dụng miễn phí (Free) hay trả phí (Paid)
* **Price**: Số tiền trả phí của ứng dụng (tại thời điểm thu thập dữ liệu)

Như vậy, ta cần kiểm tra cột `Type` có giá trị là `'Free'` tương đương với cột `Price` có giá trị là 0, cột `Type` có giá trị là `'Paid'` tương đương với cột `Price` có giá trị lớn hơn 0.

Kiểm tra các giá trị trong cột `Price` khi cột `Type` có giá trị là `'Free'`:

In [27]:
app_df[app_df['Type'] == 'Free']['Price'].unique()

array([0.])

Kiểm tra có bao nhiêu giá trị bé hơn hoặc bằng 0 trong cột `Price` khi cột `Type` có giá trị là `'Paid'`:

In [28]:
price_unique_array = app_df.loc[app_df['Type'] == 'Paid', 'Price']
(price_unique_array <= 0).sum()

0

Vậy dữ liệu trong cột `Type` và cột `Price` đã đúng theo mô tả dữ liệu.

#### Tiền xử lý cột `Content Rating`

Quan sát các giá trị khác nhau trong cột `Content Rating`:

In [29]:
np.array(sorted(app_df['Content Rating'].unique()))

array(['Adults only 18+', 'Everyone', 'Everyone 10+', 'Mature 17+',
       'Teen', 'Unrated'], dtype='<U15')

Tương tự cột `Installs`, ta thấy rằng cột `Content Rating` thuộc loại cột categorical có thứ tự, nên ta sẽ định nghĩa thứ tự các giá trị này và áp dụng cho cột `Content Rating`.

In [30]:
cr_cate_list = ['Everyone', 'Everyone 10+', 'Teen', 'Mature 17+', 'Adults only 18+']
app_df['Content Rating'] = pd.Categorical(app_df['Content Rating'], categories=cr_cate_list, ordered=True)
app_df['Content Rating']

0          Everyone
2          Everyone
3              Teen
4          Everyone
5          Everyone
            ...    
10836      Everyone
10837      Everyone
10838      Everyone
10839    Mature 17+
10840      Everyone
Name: Content Rating, Length: 9660, dtype: category
Categories (5, object): ['Everyone' < 'Everyone 10+' < 'Teen' < 'Mature 17+' < 'Adults only 18+']

#### Tiền xử lý cột `Last Updated`

Chuyển dtype của cột `Last Updated` sang kiểu dữ liệu thời gian:

In [31]:
app_df['Last Updated'] = pd.to_datetime(app_df['Last Updated'])
app_df['Last Updated']

0       2018-01-07
2       2018-08-01
3       2018-06-08
4       2018-06-20
5       2017-03-26
           ...    
10836   2017-07-25
10837   2018-07-06
10838   2017-01-20
10839   2015-01-19
10840   2018-07-25
Name: Last Updated, Length: 9660, dtype: datetime64[ns]

#### Kiểu dữ liệu hiện tại của các cột dữ liệu

In [32]:
app_df.dtypes

App                       object
Category                  object
Rating                   float64
Reviews                    int64
Size                     float64
Installs                category
Type                      object
Price                    float64
Content Rating          category
Genres                    object
Last Updated      datetime64[ns]
Current Ver               object
Android Ver               object
dtype: object

### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?

Ta có 5 cột `Rating`, `Reviews`, `Size`, `Price`, `Last Updated` thuộc nhóm các cột numerical. Với mỗi cột ta sẽ tính tỉ lệ phần trăm giá trị thiếu thuộc [0, 100], min, max.

In [33]:
def missing_ratio(s):
    return s.isna().mean() * 100

app_nume_cols = ['Rating', 'Reviews', 'Size', 'Price', 'Last Updated']
app_nume_col_profiles_df = app_df[app_nume_cols].agg([missing_ratio, pd.Series.min, pd.Series.max])
app_nume_col_profiles_df

Rating     Reviews           Size  Price  \
missing_ratio  15.144928         0.0      12.701863    0.0   
min             1.000000         0.0       8.500000    0.0   
max             5.000000  78158306.0  102400.000000  400.0   

                      Last Updated  
missing_ratio                  0.0  
min            2010-05-21 00:00:00  
max            2018-08-08 00:00:00

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

Ta có 8 cột `App`, `Category`, `Installs`, `Type`, `Content Rating`, `Genres`, `Current Ver`, `Android Ver` thuộc nhóm các cột categorical. Với mỗi cột categorical ta sẽ tính tỉ lệ phần trăm giá trị thiếu thuộc [0, 100], số lượng giá trị khác nhau (không xét giá trị thiếu), danh sách các giá trị khác nhau (không xét giá trị thiếu).

In [34]:
def num_diff_vals(s):
    return s.nunique()

def diff_vals(s):
    return s.dropna().unique()

app_cate_cols = ['App', 'Category', 'Installs', 'Type', 'Content Rating', 'Genres', 'Current Ver', 'Android Ver']
app_cate_col_profiles_df = app_df[app_cate_cols].agg([missing_ratio, num_diff_vals, diff_vals])
app_cate_col_profiles_df

App  \
missing_ratio                                                0.0   
num_diff_vals                                               9660   
diff_vals      [Photo Editor & Candy Camera & Grid & ScrapBoo...   

                                                        Category  \
missing_ratio                                           0.010352   
num_diff_vals                                                 33   
diff_vals      [ART_AND_DESIGN, AUTO_AND_VEHICLES, BEAUTY, BO...   

                                                        Installs  \
missing_ratio                                                0.0   
num_diff_vals                                                 21   
diff_vals      ['10,000+', '5,000,000+', '50,000,000+', '100,...   

                       Type  \
missing_ratio           0.0   
num_diff_vals             2   
diff_vals      [Free, Paid]   

                                                  Content Rating  \
missing_ratio                                           0.020704   
num_diff_vals                                                  5   
diff_vals      ['Everyone', 'Teen', 'Everyone 10+', 'Mature 1...   

                                                          Genres  \
missing_ratio                                           0.010352   
num_diff_vals                                                119   
diff_vals      [Art & Design, Art & Design;Creativity, Auto &...   

                                                     Current Ver  \
missing_ratio                                           0.082816   
num_diff_vals                                               2818   
diff_vals      [1.0.0, 1.2.4, Varies with device, 1.1, 1.0, 6...   

                                                     Android Ver  
missing_ratio                                           0.020704  
num_diff_vals                                                 33  
diff_vals      [4.0.3 and up, 4.2 and up, 4.4 and up, 2.3 and...

# Đưa ra các câu hỏi có ý nghĩa cần trả lời

* Câu hỏi là gì?
* Nếu trả lời được câu hỏi thì sẽ có lợi ích gì?

# Tiền xử lý + phân tích dữ liệu để trả lời cho từng câu hỏi

# Nhìn lại quá trình làm đồ án

* Mỗi thành viên: Đã gặp những khó khăn gì? (Hay mọi
chuyện đều thuận lợi)
* Mỗi thành viên: Có học được gì hữu ích? (Hay không
học được gì)
* Nhóm: Nếu có thêm thời gian thì sẽ làm gì?

# Tài liệu tham khảo